staker-rewards/lib/sla-calculator.js

```diff
@@ -0,0 +1,148 @@
+        this.deactivatedKeeps = deactivatedKeeps
+        this.signatureFailKeeps = signatureFailKeeps
+    }
+
+    static initialize(cache, interval) {
+        // Step 1 of keygen SLA: get keeps opened within the given interval
+        const openedKeeps = cache
+            .getKeeps() // get keeps with all statuses
+            .filter(keep =>
+                // get keeps whose creation timestamps are within
+                // the given interval
+                interval.start <= keep.creationTimestamp &&
+                keep.creationTimestamp < interval.end
+            )
+
+        // Step 2 of keygen SLA: from keeps opened within the given interval,
+        // get the ones which have been eventually terminated due to keygen fail
+        const keygenFailKeeps = openedKeeps
+            .filter(keep =>
+                keep.status.name === KeepStatus.TERMINATED &&
+                keep.status.cause === KeepTerminationCause.KEYGEN_FAIL
+            )
+
+        // Step 1 of signature SLA: get keeps closed within the given interval
+        const closedKeeps = cache
+            .getKeeps(KeepStatus.CLOSED) // get closed keeps
+            .filter(keep => {
+                // get keeps whose statuses have been changed within the
+                // given interval
+                return interval.start <= keep.status.timestamp &&
+                    keep.status.timestamp < interval.end
+            })
+
+        // Step 2 of signature SLA: get keeps terminated within the
+        // given interval
+        const terminatedKeeps = cache
+            .getKeeps(KeepStatus.TERMINATED) // get terminated keeps
+            .filter(keep => {
+                // get keeps whose statuses have been changed within the
+                // given interval
+                return interval.start <= keep.status.timestamp &&
+                    keep.status.timestamp < interval.end
+            })
+
+        // Step 3 of signature SLA: Concatenate keeps closed and terminated
+        // within the given interval but exclude the keeps terminated due to
+        // keygen fail as they are not relevant for the signature SLA. This way
+        // we obtain an array of keeps whose statuses have been changed from
+        // `active` to `closed`/`terminated`. Implicitly, this means a keep
+        // became not active due to one of the following causes:
+        // - keep has been closed after delivering a signature successfully
+        // - keep has been terminated after not delivering a signature
+        // - keep has been terminated from another reason not related
+        //   with the signature or keygen context
+        const deactivatedKeeps = [].concat(
+            closedKeeps,
+            terminatedKeeps.filter(
+                // get keeps which have been terminated due to causes
+                // other than keygen fail
+                keep => keep.status.cause !== KeepTerminationCause.KEYGEN_FAIL
+            )
+        )
```

Comment:
I am having second thoughts whether we should include keeps with `KeepTerminationCause.OTHER` when calculating signature SLA. Those keeps were never asked for a redemption signature and they are only increasing success rate. Please consider the following scenario: I have 101 tBTC deposits, 100 of them were heavily undercollateralized and were liquidated. I was never asked to provide a redemption signature for them. I have 1 other deposit that was terminated with `KeepTerminationCause.SIGNATURE_FAIL`. What's my SLA?

I think now that we should compute a ratio of `SIGNATURE_FAIL : (CLOSED + SIGNATURE_FAIL)`. WDYT?

Reply:
Yeah, agree. I'll get rid of the keeps with the `KeepTerminationCause.OTHER` termination cause.